### ISSUE TITLE
Watch JuliaCon in the Terminal

### ISSUE THREAD

**Issue by carstenbauer** [JuliaCon/JuliaCon.jl#14]
@logankilpatrick had an awesome idea on the JuliaCon call: Wouldn't it be super cool if one could watch JuliaCon (the YouTube streams) in the terminal? @IanButterworth, maybe [VideoInTerminal.jl](https://github.com/IanButterworth/VideoInTerminal.jl) can help with this?

**Comment by IanButterworth** [comment]
Yeah I thought the same. I wonder if we could expose a youtube stream to VideoIO

**Comment by carstenbauer** [comment]
Maybe something along the lines of https://stackoverflow.com/questions/37040798/how-do-you-use-youtube-dl-to-download-live-streams-that-are-live ? Shouldn't it be somewhat similar to how the camera stream gets displayed with VideoIO?

Unfortunately I have no experience in this area ... :(

**Comment by carstenbauer** [comment]
~~Would it be helpful if I would try to create a youtube-dl_jll?~~ (youtube-dl is a python module / requires the python interpreter)

**Comment by IanButterworth** [comment]
This was easier than I expected... 
For the live videos we'd want to pipe the frames directly into julia, rather than download to file like this though


```julia
using HTTP, JSON3, FFMPEG, VideoInTerminal

function get_youtube(id = "j18ECUhkeY0")
    f, io = mktemp()
    dest = string(f,".mp4")

    r = HTTP.request("GET", "https://www.youtube.com/get_video_info?video_id=$(id)&html5=1")
    lines = split(HTTP.unescapeuri(String(r.body)), '&')
    for line in lines
        if startswith(line, "player_response")
            pr = JSON3.read(split(line, "player_response=", keepempty = false)[1])
            for format in pr[:streamingData][:adaptiveFormats]
                if format[:height] == 240
                    FFMPEG.exe("-i", format[:url], dest, collect=true)
                    return dest
                end
            end
        end
    end
    error("Could not find usable video")
end

function watch_juliacon(id = "TPuJsgyu87U")
    @info "Grabbing video..."
    vid = get_youtube(id)
    VideoInTerminal.play(vid)
end
```
<img width="763" alt="Screen Shot 2021-06-05 at 12 38 46 AM" src="https://user-images.githubusercontent.com/1694067/120880169-96c21480-c596-11eb-8d73-3e1e08794b1e.png">

<img width="767" alt="Screen Shot 2021-06-05 at 12 39 16 AM" src="https://user-images.githubusercontent.com/1694067/120880168-96297e00-c596-11eb-99ba-3e8a1ac0f3df.png">

name that talk...

**Comment by carstenbauer** [comment]
This is awesome! Needs a little bit of tweaking though to handle live streams I guess? Example: `id = 21X5lGlDOfg` (https://www.youtube.com/watch?v=21X5lGlDOfg) doesn't work.

**Comment by carstenbauer** [comment]
And, just for my understanding, we're still downloading to file in some sense right? (`FFMPEG.exe("-i", format[:url], dest, collect=true)`

**Comment by carstenbauer** [comment]
Sorry, one more thing. Any idea how to make audio work ? 😀

**Comment by IanButterworth** [comment]
> Sorry, one more thing. Any idea how to make audio work ? 

Nope. It might be possible to extract a subtitle file and print that underneath though 

For live streams I think it would probably have to be through VideoIO via an AVInput, but I couldn't figure it out quickly

**Comment by oscardssmith** [comment]
If we want maximum style points, we could use the `say` command

**Comment by Moelf** [comment]
I can probably add a high-definition version for `Kitty` terminal if the interests is there....

https://user-images.githubusercontent.com/5306213/119249864-1baa3880-bb6a-11eb-90fe-c7cca48692f1.mp4

**Comment by carstenbauer** [comment]
I'd still love to have this... it's so cool :) Maybe we could try to use https://github.com/eschnett/SixelTerm.jl to get a reasonable quality? :)
